File: 3.1/inflate/inflate.py
"""
ID: giliev91
LANG: PYTHON3
TASK: inflate
"""


def input():
    with open("inflate.in", "r") as f:
        M, _N = tuple(map(int, f.readline().strip().split()))
        categories = [tuple(map(int, line.strip().split()))
                      for line in f.readlines()]

        return M, categories


def solve(M, categories):
    best = [0] * (M+1)
    for i in range(M+1):
        for points, time in categories:
            if i >= time:
                best[i] = max(best[i], best[i-time] + points)

    return best


def output(result):
    with open("inflate.out", "w") as f:
        f.write("%d\n" % result)


M, categories = input()
best = solve(M, categories)
output(best[M])

File: 3.1/inflate/inflate.cpp
/*
ID: giliev91
LANG: C++
TASK: inflate
*/
#include <iostream>
#include <fstream>
using namespace std;

int main()
{
    ifstream fin("inflate.in");
    ofstream fout("inflate.out");

    int M, N;
    fin >> M >> N;

    int *points = new int[N];
    int *time = new int[N];
    for (int i = 0; i < N; i++)
    {
        fin >> points[i] >> time[i];
    }

    int *best = new int[M + 1];
    fill(best, best + sizeof(best), 0);
    for (int i = 0; i <= M; i++)
        for (int j = 0; j < N; j++)
            if (i >= time[j])
                if (best[i] < best[i - time[j]] + points[j])
                    best[i] = best[i - time[j]] + points[j];

    fout << best[M] << "\n";
    return 0;
}